src/java/org/apache/cassandra/cql3/statements/GrantPermissionsStatement.java

```diff
@@ -39,7 +43,25 @@ public GrantPermissionsStatement(Set<Permission> permissions, IResource resource
 
     public ResultMessage execute(ClientState state) throws RequestValidationException, RequestExecutionException
     {
-        DatabaseDescriptor.getAuthorizer().grant(state.getUser(), permissions, resource, grantee);
+        IAuthorizer authorizer = DatabaseDescriptor.getAuthorizer();
+        Set<Permission> granted = authorizer.grant(state.getUser(), permissions, resource, grantee);
+
+        // We want to warn the client if all the specified permissions have not been granted and the client did
+        // not specified ALL in the query.
+        if (!granted.equals(permissions) && !permissions.equals(Permission.ALL))
+        {
+            // We use a TreeSet to guarantee the order for testing
+            String permissionsStr = new TreeSet<>(permissions).stream()
+                                                              .filter(permission -> !granted.contains(permission))
+                                                              .map(Permission::name)
+                                                              .collect(Collectors.joining(", "));
```

Comment:
Given that we are creating a stream, we could save us the `TreeSet` by just ordering the stream, for example:
```suggestion
            String permissionsStr = permissions.stream()
                                               .filter(permission -> !granted.contains(permission))
                                               .sorted(Permission::compareTo) // guarantee the order for testing
                                               .map(Permission::name)
                                               .collect(Collectors.joining(", "));
```